code/modules/holodeck/computer.dm

```diff
@@ -110,63 +120,193 @@
-				if(P["type"] == program_to_load)
-					valid = TRUE
-					break
-			if(!valid)
-				return FALSE
-
-			var/area/A = locate(program_to_load) in GLOB.sortedAreas
-			if(A)
-				load_program(A)
+			var/program_to_load = params["id"]
+			//load the map_template that program_to_load represents
+			if(program_to_load)
+				load_program(program_to_load)
 		if("safety")
 			if((obj_flags & EMAGGED) && program)
 				emergency_shutdown()
 			nerf(obj_flags & EMAGGED)
 			obj_flags ^= EMAGGED
 			say("Safeties restored. Restarting...")
 
+/datum/map_template/holodeck/update_blacklist(turf/placement)//this is what makes the holodeck not spawn anything on broken tiles (space and non engine plating)
+	turf_blacklist.Cut()
+	for (var/_turf in get_affected_turfs(placement))
+		var/turf/possible_blacklist = _turf
+		if (!istype(possible_blacklist, /turf/open/floor/holofloor))
+			if (istype(possible_blacklist, /turf/open/floor/engine))
+				continue
+			turf_blacklist += possible_blacklist
+
+///the main engine of the holodeck, it loads the template whose id string it was given ("offline_program" loads datum/map_template/holodeck/offline)
+/obj/machinery/computer/holodeck/proc/load_program(map_id, force = FALSE, add_delay = TRUE)
+	if (program == map_id)
+		return
+
+	if (!is_operational)//load_program is called once with a timer (in toggle_power) we dont want this to load anything if its off
+		map_id = offline_program
+		force = TRUE
+
+	if (current_cd > world.time && !force)
+		say("ERROR. Recalibrating projection apparatus.")
+		return
+
+	if (add_delay)
+		current_cd = world.time + HOLODECK_CD
+		if(damaged)
+			current_cd += HOLODECK_DMG_CD
+	active = (map_id != offline_program)
+	use_power = active + IDLE_POWER_USE
+	program = map_id
+
+	//clear the items from the previous program
+	if (spawned)
+		for (var/_item in spawned)
+			var/obj/holo_item = _item
+			derez(holo_item)
+	if (effects)
+		for (var/_effect in effects)
+			var/obj/effect/holodeck_effect/holo_effect = _effect
+			effects -= holo_effect
+			holo_effect.deactivate(src)
+
+	template = SSmapping.holodeck_templates[map_id]
+	for (var/_turf in linked)
+		var/turf/holo_turf = _turf
+		if (istype(holo_turf, /turf/closed))
+			for (var/_baseturf in holo_turf.baseturfs)
+				if (ispath(_baseturf, /turf/open/floor/holofloor))
+					holo_turf.baseturfs -= _baseturf
+					holo_turf.baseturfs += /turf/open/floor/holofloor/plating
+
+	template.load(bottom_left)//this is what actually loads the holodeck simulation into the map
+
+	spawned = template.created_atoms
+
+	nerf(!(obj_flags & EMAGGED))
+	finish_spawn()
+
+/obj/machinery/computer/holodeck/proc/finish_spawn()//this is used for holodeck effects (like spawners). otherwise they dont do shit
+	for (var/_atom in spawned)
+		var/atom/atoms = _atom
+
+		if (isturf(atoms))
+			var/turf/holo_turf = atoms
+			spawned -= holo_turf
+
+		atoms.flags_1 |= HOLOGRAM_1
+
+		if (istype(atoms, /obj/effect/holodeck_effect/))//this is what makes holoeffects work
```

Comment:
Would be nice to wrap these holoeffects into some typed proc while you're doing this.